enhancements/cvo/upgrade_preflight_checks.md

```diff
@@ -0,0 +1,107 @@
+
+* How CVO can get the list of operators which has the preflight checks implemented?
+* What information CVO needs from the operators to create the preflight jobs?
+* Does CVO already has capability to create jobs for preflight checks?
+* How to trigger execution of preflight checks?
+* How CVO will report when it gets the result from preflight checks?
+* What information CVO needs from operators to create the preflight job?
+* How will preflight checks report back success or failure, and the detailed info for administrators?
+
+## Summary
+
+The goal of over the air update(OTA) feature of OpenShift is to make upgrades seamless and make the upgrade a success almost always.
+To improve the upgrade success we should be able to detect if the new version is compatible with the existing version before running the actual upgrade.
+As OpenShift V4 is composed of many operators, we need to give operators ability to run some checks to find out if the new version is compatible with the existing version and use the result to improve the success rate of upgrades.
+
+## Motivation
+
+CVO managed operators should be able to run preflight checks before the actual upgrade.
+It would help operators to find out if they can move to the specific version safely before doing an actual upgrade.
+Also administrators are empowered to run these preflight checks without doing an upgrade.
+
+
+### Goals
+
+* Administrators will be able to run preflight checks without running the upgrade.
+* Give Operators (CVO managed) ability to add checks to check if they can safely upgrade to a given release.
+* Preflight checks would be idempotent in nature.
+
+### Non-Goals
+
+* This enhancement will not add code required for preflight checks for individual operators.
+* Operators will have to figure out the actual code they would like to run as part of preflight checks.
+* The preflight checks should not be used as a migration mechanism.
+
+## Proposal
+
+Operators need to provide a different bin path for the preflight checks i.e. /bin/preflight.
+The binpath need to exist inside of the operator image.
+Cluster version operator (CVO) will be running a single job per operator for the preflight checks.
+The binpath /bin/preflight must be executable and it can be symbolic link to the operator binary or a script or It can be non existent.
+CVO will not fail if the path does not exist in an operator image but it would assume that preflight checks are not implemented for the respective operator.
+Also there would be a possible way to skip the preflight checks if someone wants to skip it explicitly.
+
+### User Stories
+
+#### Story 1
+
+Some customers would prefer to disable automatic credential provisioning by Cloud credential operator (CCO) and instead manage it themselves.
+However situations can arise during an upgrade where credentials have changed and the upgrade will be block and go into a failed state, with little visibility as to why or how it can be fixed.
```

Comment:
I think the visibility issue is fixable and orthogonal.  The main issue with a lack of preflights is that if you hang mid-update while you sort out these cred issues, the CVO is no longer applying or monitoring the later-in-graph manifests to stomp changes that cluster admins might be making.  I guess we could fix that too, but trying to figure out which source-release manifests should still be reconciled when you're at a particular point in the target-release manifest graph seems complicated.  But getting back to this sentence, can we reword to something like:

> If new permissions are required by the target version, the update could block midway while the admin provisioned more powerful credentials.

and punt on further explainations of why a blocked update is bad?

Reply:
Fixed https://github.com/openshift/enhancements/commit/190ec1f0c51076d5e5b5b6b70dc7911db2fc51cc